torchmultimodal/modules/layers/transformer.py

```diff
@@ -328,11 +341,71 @@ def forward(
             )
 
 
-def _apply_layernorm(x: Tensor, layernorm: nn.Module) -> Tensor:
-    """Supports mixed-precision training by casting to fp32 for layernorm and back"""
-    if x.dtype != torch.float32:
-        x_fp32 = x.float()
-        x_fp32 = layernorm(x_fp32)
-        return x_fp32.type_as(x)
-    else:
-        return layernorm(x)
+class TransformerEncoder(nn.Module):
+    def __init__(
+        self,
+        encoder_layer: nn.Module,
+        num_layers: int,
+    ):
+        super().__init__()
+        self.layer = get_clones(encoder_layer, num_layers)
+        self.num_layers = num_layers
+
+    def forward(
+        self,
+        hidden_states: Tensor,
+        attention_mask: Optional[Tensor] = None,
+        head_mask: Optional[Tensor] = None,
+        return_attn_weights: bool = False,
+        return_hidden_states: bool = False,
+    ) -> TransformerOutput:
+        all_hidden_states: Tuple[Tensor, ...] = () if return_hidden_states else None
+        all_self_attentions: Tuple[Tensor, ...] = () if return_attn_weights else None
+
+        for i, layer_module in enumerate(self.layer):
+            if return_hidden_states:
+                all_hidden_states = all_hidden_states + (hidden_states,)
+
+            layer_head_mask = head_mask[i] if head_mask is not None else None
+            layer_outputs = layer_module(
+                hidden_states,
+                attention_mask=attention_mask,
+                head_mask=layer_head_mask,
+                return_attn_weights=True,
+            )
+
+            hidden_states = layer_outputs[0]
```

Comment:
If `layer_outputs` only contains the hidden_states, this will try to index the 0-th element instead of returning the full hidden_states, right?

```
        if return_attn_weights:
            return outputs, attn_weights
        else:
            return outputs
```

Consider adding a unit test when the kwargs args are False as well?

```
output = encoder(inputs, return_hidden_states=True, return_attn_weights=True)
```

Reply:
Not sure if this has been addressed as the encoder layer doesn't always return two tensors, right?

https://github.com/facebookresearch/multimodal/blob/adc36c18c0ccae7652eee34256ecc74e628f7aba/torchmultimodal/modules/layers/transformer.py#L381

Reply:
return_attn_weights is set to True so it will guarantee returning two outputs

Reply:
Not sure. Then we should probably ask `TransformerEncoderLayer` to always return two outputs anyways.

Reply:
hmm, maybe I'll modify to handle both single and double outputs